### ISSUE TITLE
Pipelining bugs in Twisted's SMTP server

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8126]
|<img alt="al45tair's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| al45tair reported|
|-|-|
|Trac ID|trac#8126|
|Type|defect|
|Created|2015-11-25 16:38:41Z|

`twisted.mail.smtp` suffers from a few problems, namely:

* If the `IMessageDelivery` implementation returns a `Deferred`, the server continues accepting commands while the deferred processing takes place. If a bad client implementation tries to send a command during this period, the exact behaviour may be unclear — but very likely the response to the new command will be sent before the response to e.g. the MAIL FROM: command that is still being processed. **This is a bug**
* It’s difficult to support `PIPELINING` in an `ESMTP` subclass — there are no hooks in the `SMTP` implementation that could be used for that purpose.
* It's impossible in an `ESMTP` subclass to support `CHUNKED` or `BINARYMIME` because `SMTP` is a subclass of `LineOnlyReceiver`.
* While the code in the `SMTP` class does account for `ESMTP` options on `MAIL FROM:` and `RCPT TO:` commands, there is no attempt to parse them or indeed to remember the options that have been specified. This makes it hard to implement ESMTP extensions in subclasses of `ESMTP` because it forces an implementor to override `do_MAIL` and `do_RCPT`.

In general, it’s desirable to detect things like attempts to send commands prior to the greeting, attempts to pipeline when pipelining is not supported, attempts to pipeline before the `EHLO` response when it _is_ supported, and so on, and reject them. Why? Because poorly written spamming software will do all of the above, while real mail server software will not.

An example of the first problem:

```
220 localhost NO UCE NO UBE NO RELAY PROBES ESMTP
EHLO example.com
250 localhost Hello 127.0.0.1, nice to meet you
MAIL FROM: test@example.com
MAIL FROM: foo@bar.com
550 blacklisted
250 Sender address accepted
```

In this case, the `validateFrom` method is looking the sender address up in a database. `foo@bar.com` is in the blacklist table in this database, hence `550 blacklisted`, but _note that that is the **first** response_, whereas the `MAIL FROM: foo@bar.com` command is the second command.

Anyway, the fix for this problem is, whenever `SMTP` hands control to a `Deferred`, to change to a new mode, so that we can detect that there has been an attempt to send a command while an existing command is still being processed. In the case of the `SMTP` class, the default response should be to give a 503 error and then close the connection — though the latter step should be configurable.

In the case of the `ESMTP` class, in the case where the client has sent `EHLO` and pipelining is enabled, rather than changing mode the code should use `pauseProducing` and `resumeProducing` to block the reception of further commands. Where pipelining is not enabled or the client has sent `HELO`, it should use the same behaviour as the base class.

Attachments:

* [smtp-pipelining-8126-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c75/c757bd113039484d4766f5ddda49f59c4c695005/de8302add39fd257be5c96814affb265047c2b6d.patch) (17890 bytes) - added by al45tair on 2015-11-25 16:40:40Z - Fixes pipelining-related bug, implements PIPELINING (off by default) for ESMTP
* [smtp-pipelining-8126-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c75/c757bd113039484d4766f5ddda49f59c4c695005/3e67956fc2de6eea5e3d907956a8861fd08a8a89.patch) (16596 bytes) - added by al45tair on 2015-11-26 15:38:47Z - Stripped out the ESMTP option processing part of the patch. Added some docstrings & some code formatting changes.

<details><summary>Searchable metadata</summary>

```
trac-id__8126 8126
type__defect defect
reporter__al45tair al45tair
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__new new
resolution__None None
component__mail mail
keywords__None None
time__1448469521538682 1448469521538682
changetime__1448575673222777 1448575673222777
version__None None
owner__al45tair al45tair

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="al45tair's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|al45tair commented|
|-|-|

I've attached a patch that fixes all of the above issues. The reason for implementing `PIPELINING` at the same time as fixing the bug is that the bug is closely related to pipelining and we may as well add it. The patch does not add support for `CHUNKED` or `BINARYMIME` or any such thing, but it does also fix the obstacles that make it hard to do so (by making `SMTP` derive from `LineReceiver` and by having it store the options it is passed in `MAIL FROM:` and `RCPT TO:` commands, so that an extension implementor need not re-implement `do_MAIL` and `do_RCPT`).

The default policy implemented by the patch when a serious protocol error is detected (such as talking before the greeting, or trying to use pipelining when it either isn't enabled or hasn't yet been advertised) is to drop the connection. This is controlled by the `protocolViolation` method on the `SMTP` class. It’s possible that e.g. a tarpit implementation might want to do something else instead — for example, deliberately start communicating very slowly in order to waste client resources, which is why all such disconnection goes through this one method.

The patch includes some new tests.

Note that I haven’t made any attempt to support Python 3.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to al45tair|
|-|-|

Many thanks for your bug report and the patch.

This is just a quick, high level review.

Your patched adds a lot of public methods without any docstring, narrative documentation or examples.

I know that the current code is not state of the art, but this should not prevent us from writing better code.

If there are multiple problems in the SMTP implementation please try to create a ticket + patch for each independent issues.

This will help us in following up the progress for each change in part and merge a fix as soon as it is ready, rather than waiting for all changed to be accepted.

----------

The patch also needs a news file fragment so that all these great changes that you have made are advertised in the release notes.

More info here http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

------

Twisted mail is not ported on Py3 so there is no need to care about Py3 for now... but it would be nice to write code which later would make the porting easier :)

--------

I have not check the test coverage. Just an info. The new patches should include full coverage for all the changes :)

For writing test docstrings you can check this article https://jml.io/pages/test-docstrings.html


The test docstring should also help why that behavior is required.. not just what it does. 
During a refactor this should help to evaluate whether a test needs to also be updated or the code needs to be updated to match the existing tests. 

So instead of

```
Test that trying to send pipelined commands to the SMTP server
causes it to error and disconnect.
```

Maybe we can have:

```
Pipelining in not supported in STMP and any pipelined request will trigger
an error and disconnection as specified in RFC 1234 section 5.
```

-------

Please check our code standards https://twistedmatrix.com/documents/current/core/development/policy/index.html

2 empty lines between methods and 3 empty lines between classes are the norm :)

---------

Please consider splitting the patch + writing docstrings. For the new features, if there a any please consider writing narrative documentation and examples.

I know that you have described your changes in this ticket, but when people are reading the code or trying to read the code they will not search the Trac ticket history to find info about a specific piece of code.

A server and client example which could help with manual testing and validation would be of great help during the review process and later can server as documentation for people who want to debug, report bugs or implement these features.

---------

Basically a perfect patch don't require any extra info the the Trac comments.

The newfile fragment should describe the scope of the patch and the docstring + narrative documentation should describe what it can do, what it can't do and how to use the API :)

---------

Please note that I am not an STMP or ESMTP expert. 
If you think that my comments have no sense, feel free to put your patch back for review and ask a review from someone else.

Many thanks again for your contribution!
Looking forward for the next patches :)

**Comment by twisted-trac** [comment]
|<img alt="al45tair's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|al45tair commented|
|-|-|

Thanks for the review comments.

Regarding docstrings (and indeed the empty lines between methods/classes), I was really copying the style of the existing code as that's normally a better policy than changing things. However, it's your project, so I will do as you ask.

As far as the multiple problems, multiple tickets thing, I did wonder if you’d say that. I've split off [#8128](https://github.com/twisted/twisted/issues/8128) containing the option parsing and `LineReceiver` change, and will update this ticket later to remove those changes from the patch.

BTW, I’m not sure why, but my user account on Trac appeared to have been deleted overnight. Maybe a server problem?

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

Thanks for the feedback.

I don't see how not adding docstring ( and narrative documentation or examples) just because the existing code does not have them, is a better policy :)

It is not my project, but rather a common good ... and I want to merge code which will generate as few WTF/m as possible to the next person who will have to maintain or work with the merged code :)

I am not the only person who can review this code, so if you think that I am asking something wrong, please resubmit and I will try to find someone else to review the patch.

Looking forward for the next patch :)


----------

There are known problems with Trac users db. Feel free to join #twisted-dev or #twisted-admin if you get into serious troubles with Trac.

**Comment by twisted-trac** [comment]
|<img alt="al45tair's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|al45tair commented|
|-|-|

No, I don't think you're asking anything wrong. Some projects prefer that new contributions match the style of the surrounding code, rather than trying to follow any coding style that has been imposed since then, that’s all.

I've updated the patch a bit, adding docstrings, reformatting a little and splitting off the option parsing into [#8128](https://github.com/twisted/twisted/issues/8128).

I haven’t added documentation or additional examples — the entire `twisted.mail` subproject is badly in need of new documentation and I think perhaps the right thing is to open a separate ticket for that and handle it all in one go.

As far as testing goes, the new test cases should cover the new code. The docstrings for the test cases are unchanged as yet; I don’t think the RFCs really specify what should happen if you don’t wait for responses from the server before sending another command, other than disallowing it — RFC5321 says simply that

   The dialog is purposely lock-step, one-at-a-time,
   although this can be modified by mutually agreed upon extension
   requests such as command pipelining (RFC 2920 [19]).

The main point is that spamming software does do these kind of things, and with the present implementation the server could get itself into all kinds of odd states by processing commands concurrently; depending on users’ code, this could even result in a security hole. Implementing the code to prevent this also makes it easy to add `PIPELINING` support to the `ESMTP` class at the same time, as doing so really just means a slight change of behaviour in the `beginProcessing` and `endProcessing` methods I added, and that’s why that’s part of this same patch.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

I agree that twisted.mail is in need for documentation, but we need to start from somewhere... just delaying the moment when we start writing good documentation while continuing to add badly documented code would not solve it.

For example we can stop accepting changes if they don't come with tests, narrative docs and examples which domenstrate how to use them.

If you also think that twisted.mail is badly documented, let's try writing new code which breaks this bad habit and comes with nice documentation and examples :)

----------

Breaking this patch into smaller one is great :)

----------

Just asking: Rather than adding new API, can we just overwrite/implement custom pauseProducing/resumeProducing method to accomodate the required functionalities ?

-----

The reference to RFC in the test docstring was just an example. In case it is not coverted by RFC we sould describe the behaviour in the docstring

------

This patch adds a lot of public methods which are not documented.

Do we really need all these public APIs ? If yes, please document why and add some examples of external code using the API :)

A few examples of new public API

```
twisted.mail.stmp.PROCESSING 
twisted.mail.smtp.STMP.deferGreeting
twisted.mail.smtp.STMP.saved_mode ... which should be savedMode
twisted.mail.smtp.ESTMP.pipelining
```

Public API is very important as once we introduce a new API we commit to keep it available for some time and only remove it following the deprecation policy.

You can find out more about our deprecation policy here https://github.com/twisted/twisted/compare/trunk...deprecation-docs-8082 

That document will soon be merged.

---------

Please check my comments and resubmit a new patch.

I find it hard to accept this patch without proper docstrings and documentation.

Thanks!
